pkg/controller/replica.go

```diff
@@ -0,0 +1,335 @@
+			cancelFunc := v.(context.CancelFunc)
+			cancelFunc()
+			delete(r.cleanUpMap.mapping, volumeName)
+		}
+		r.cleanUpMap.unlock()
+	}
+}
+
+func (r *ReconcileReplica) manageReplicas(ctx context.Context, volumeName string) error {
+	// this is to prevent multiple sync volume operation to be performed on a single volume concurrently as it can create or delete more attachments than necessary
+	v, ok := r.mutexMap.get(volumeName)
+	if ok {
+		volMutex := v.(*sync.RWMutex)
+		volMutex.Lock()
+		defer volMutex.Unlock()
+	}
+
+	azVolume, err := azureutils.GetAzVolume(ctx, r.client, r.azVolumeClient, volumeName, r.namespace, true)
+	if errors.IsNotFound(err) {
+		klog.Infof("Aborting replica management... volume (%s) does not exist", volumeName)
+		return nil
+	} else if err != nil {
+		klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
+		return err
+	}
+	if !isCreated(azVolume) {
+		return status.Errorf(codes.Aborted, "azVolume (%s) has no underlying volume object", azVolume.Name)
+	}
+
+	// get all replica attachments for the given volume
+	volReq, err := createLabelRequirements(azureutils.VolumeNameLabel, volumeName)
+	if err != nil {
+		return err
+	}
+	roleReq, err := createLabelRequirements(azureutils.RoleLabel, string(v1alpha1.ReplicaRole))
+	if err != nil {
+		return err
+	}
+	labelSelector := labels.NewSelector().Add(*volReq, *roleReq)
+	azVolumeAttachments := &v1alpha1.AzVolumeAttachmentList{}
+	err = r.client.List(ctx, azVolumeAttachments, &client.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		klog.Errorf("failed to list AzVolumeAttachment: %v", err)
+		return err
+	}
+
+	desiredReplicaCount, currentReplicaCount := azVolume.Spec.MaxMountReplicaCount, len(azVolumeAttachments.Items)
+	klog.Infof("control number of replicas for volume (%s): desired=%d,\tcurrent:%d", azVolume.Spec.UnderlyingVolume, desiredReplicaCount, currentReplicaCount)
+
+	// if the azVolume is marked deleted, do no create more azvolumeattachment objects
+	if azVolume.DeletionTimestamp == nil && desiredReplicaCount > currentReplicaCount {
+		klog.Infof("Need %d more replicas for volume (%s)", desiredReplicaCount-currentReplicaCount, azVolume.Spec.UnderlyingVolume)
+		if azVolume.Status.Detail == nil || azVolume.Status.State == v1alpha1.VolumeDeleting || azVolume.Status.State == v1alpha1.VolumeDeleted || azVolume.Status.Detail.ResponseObject == nil {
+			// underlying volume does not exist, so volume attachment cannot be made
+			return nil
+		}
+		if err = r.createReplicas(ctx, min(defaultMaxReplicaUpdateCount, desiredReplicaCount-currentReplicaCount), azVolume.Name, azVolume.Status.Detail.ResponseObject.VolumeID); err != nil {
+			klog.Errorf("failed to create %d replicas for volume (%s): %v", desiredReplicaCount-currentReplicaCount, azVolume.Spec.UnderlyingVolume, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileReplica) getNodesForReplica(ctx context.Context, volumeName string, numReplica int) ([]string, error) {
```

Comment:
Wouldn't this lead to concurrency issues? Let's say that a pod uses 2 volumes, and the volumes have primary attachment on node1 and replicas on node 2. Node 1 fails, and the volume attachments on node 2 should be converted to primary. At this point we reconcile and try to create replica attachments for volume 1 and volume 2. We might get reconciliation requests for these 2 concurrently and end up selecting 2 different nodes.

Reply:
Yeah, good point. We might need a pod-level mutex to enforce single replica reconciliation for each pod at a time.